### PR TITLE
Remove shap as core dependency

### LINKED ISSUE #7681: [BUG] Can't install on python 3.11
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 2.1.1
- Tracking server: n/a

### System information

- macOS 13.1:
- Python Version: 3.11.1:

### Describe the problem

Can't install mlflow because numba is required which depends on llvmlite which is not yet python 3.11 compatible (https://github.com/numba/numba/issues/8304) (Can this numba dependency be optional for the time being?)
The error is that llvmlite explicitly mentions that it needs python <3.11 when installing mlflow.

### What language(s) does this bug affect?

- [x] `language/python`: Python APIs and clients

### PR DESCRIPTION
**Pull request by jmahlik**
## Related Issues/PRs

Closes #7681

## What changes are proposed in this pull request?

Moving shap to an extra and gating the shap tab in recipes behind shap being installed. This was the one place I could find with a hard dependency on shap. Happy to make other adjustments if there are places that were missed.

This likely can be classed as a bug fix. I'm not sure if it would be considered a breaking change, since users will have to install mlflow and manage the shap install separately. It should for sure be mentioned in the release notes. 

## How is this patch tested?
Tested most of the relevant functionality without shap installed on 3.11. It's quite hard to run the full test suite on Windows/3.11 since some of the extra ml requirements are not readily installable.

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Move `shap` to an extra ml requirement to improve python 3.11 compatibility. Shap needs to be installed separately for the shap functionality in autologging/recipies.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by jmahlik** [comment]
Landed on installing shap for the tests vs. removing it entirely and checking the for error logs. That seems to make more sense.

File: requirements/test-requirements.txt [hunk 1]
```diff
@@ -21,7 +21,9 @@ ipython
 # 1.1.0 contains this issue: https://github.com/microsoft/FLAML/issues/871
 flaml!=1.1.0
 # Required by custom flavor example
-sktime==0.16.0
+sktime>=0.16.0
```
Comment (jmahlik) [inline requirements/test-requirements.txt:24]: Needed to bump to run on 3.11, can repin if needed.

Comment (BenWilson2) [reply requirements/test-requirements.txt:24]: I think this is fine

**Review by BenWilson2 — APPROVED** [review]
LGTM! Thanks for the contribution @jmahlik :)

**Comment by BenWilson2** [comment]
@sunishsheth2009 any objections to having a card not populated so that we don't have to have shap as a hard dependency?

**Comment by sunishsheth2009** [comment]
> @sunishsheth2009 any objections to having a card not populated so that we don't have to have shap as a hard dependency?

I think this makes sense. Maybe we should add it here in that case: https://github.com/mlflow/recipes-examples/blob/main/requirements.txt

**Comment by jmahlik** [comment]
Rebased to resolve the conflict.

**Comment by jmahlik** [comment]
Looks like some possibly flakey tests failing for 404's on installing pyspark and mlleap serialization.

**Comment by BenWilson2** [comment]
@jmahlik those test failures were due to the release of PySpark3.4 and a recent breaking release of mlserver. They've been fixed in master. :)

**Comment by jmahlik** [comment]
Rebased.

**Comment by BenWilson2** [comment]
Thanks for another great contribution @jmahlik 🎉

**Pull request merged**